File: application/service/Decorate/Params/Equipment/Weapon.py
from models.Equipment.Weapon import Common

import helpers.mongo

from models.User.Factory import User_Factory


class Decorate(object):
    def _fixParams(self, data):
        assert(data['type'] in [Common.TYPE_BOW, Common.TYPE_SPEAR, Common.TYPE_BLUNT, Common.TYPE_SWORD])

        if 'user' in data:
            data['user'] = helpers.mongo.objectId(data['user'])

        data['damage'] = int(data['damage'])
        data['speed'] = int(data['speed'])
        data['critical_chance'] = int(data['critical_chance'])
        data['critical_damage'] = float(data['critical_damage'])

        return data

    def save(self, data, user=None):
        data = self._fixParams(data)
        return super().save(data, user)

    def simulate(self, data):
        data = self._fixParams(data)
        return super().simulate(data)

    def get(self, _id, user=None):
        _id = helpers.mongo.objectId(_id)
        return super().get(_id, user)

    def load(self, user):
        userId = helpers.mongo.objectId(user)
        user = User_Factory.getDomainById(userId)
        return super().load(user)

File: public/js/model/equipment/weapon.js
define('model/equipment/weapon', [
], function () {
    return AbstractModel.extend({
        model_url: "equipment/weapon",

        load: function (successCallback) {
            this.sync('get', {
                data: {
                    _id: this.get('_id')
                },
                success: successCallback
            });
        },

        save: function (successCallback) {
            this.sync('save', {
                data: {
                    damage: this.get('damage'),
                    speed: this.get('speed'),
                    critical_damage: this.get('critical_damage'),
                    critical_chance: this.get('critical_chance'),
                    type: this.get('type'),
                    user: this.get('user')
                },
                success: successCallback
            });
        },

        simulate: function () {
            if (
                !this.get('damage') ||
                !this.get('speed') ||
                !this.get('critical_damage') ||
                !this.get('critical_chance') ||
                !this.get('type')
            ) {
                return;
            }

            this.sync('simulate', {
                data: this.attributes
            });
        },

        remove: function (success) {
            this.sync('remove', {
                data: {
                    '_id': this.get('_id')
                },
                success: success
            });
        }
    });
});